nomad/consul_policy.go

```diff
@@ -67,17 +67,43 @@ func (c *consulACLsAPI) isManagementToken(token *api.ACLToken) bool {
 	return false
 }
 
-// namespaceCheck is used to verify the namespace of the object matches the
-// namespace of the ACL token provided.
+// namespaceCheck is used to fail the request if the namespace of the object does
+// not match the namespace of the ACL token provided.
 //
-// exception: iff token is in the default namespace, it may contain policies
+// *exception*: if token is in the default namespace, it may contain policies
 // that extend into other namespaces using namespace_prefix, which must bypass
 // this early check and validate in the service/keystore helpers
+//
+// *exception*: if token is not in a namespace, consul namespaces are not enabled
+// and there is nothing to validate
+//
+// If the namespaces match, whether the token is allowed to perform an operation
+// is checked later.
 func namespaceCheck(namespace string, token *api.ACLToken) error {
-	if token.Namespace != "default" && token.Namespace != namespace {
+
+	switch {
+	case namespace == "" && token.Namespace == "":
+		// ACLs not enabled (if Consul ACLs are enabled, the token response from
+		// consul will always set the namespace (to "default" if otherwise unset))
+		return nil
+
+	case token.Namespace == "default":
+		// ACLs enabled, must defer to per-object checking, since the token could
+		// have namespace_prefix blocks allowing the operation
+		return nil
+
+	case namespace == token.Namespace:
+		// ACLs enabled, namespaces are the same
+		return nil
+
+	case namespace == "" && token.Namespace != "default":
+		// ACLs enabled with non-default token, but namespace on job not set, so
+		// provide a more informative error message.
+		return errors.Errorf("consul ACL token requires using namespace %q", token.Namespace)
+
+	default:
```

Comment:
I found it a tad harder to reason through conditions, maybe because of ordering. I would probably find it easier if the conditions were order the following way:

1. `namespace == token.Namespace`: either ACL not enabled, or it's enabled and they match
2. `token.Namespace == "default"`: special case
3. `token.Namespace != "default" && namespace == ""` <- `token.Namespace != "default"` is redundant here but included for explicitness
4. `default`: the mismatch case

Reply:
Thanks for this suggestion, this is easier to read :slightly_smiling_face: